### PR TITLE
Add flag to filter steps which must be executed

### LINKED ISSUE #363: Filter steps which should be executed
**Problem.** Unable to choose steps which 'Universum' should execute. Alternative is a separate 'config' files usage, however it's not user-friendly enough.

**Supposed use-case.** 
A 'Universum config' contain a lot of steps related to build project for a different platforms and/or configurations. 
For example, 'config' contain next steps: 'Build Android', 'Build TEEgris', 'Build Tizen', 'Run Unit tests', 'Test Android'. It could be useful to run only **'Android'** steps.
```bash
universum [args] --filter "*Android*"
```
In this case only 'Build Android' and 'Test Android' steps should be executed.

Also, it could be useful to apply multiple filters:
```bash
universum [args] --filter "*Android*:*test*"
```
In this case the following steps will be executed: 'Build Android', 'Test Android', 'Run Unit tests'.

This feature will highly improve user experience for projects which use 'Universum' as build system wrapper.

### PR DESCRIPTION
**Pull request by i-savynikh**
Add flag '--run-step' to filter steps which should be executed.

Resolves #363

### CODE REVIEW

File: _universum/modules/launcher.py [hunk 1]
```diff
@@ -85,6 +85,75 @@ def check_if_env_set(configuration):
     return True
 
 
+def check_str_match(string, expected_substrings, unexpected_substrings):
+    """The function to check whenever specified string contains 'expected' and
+    does NOT contains 'unexpected' substrings.
+
+    ["", ["parent 1", "parent 2", "parent 1 step 1", "parent 2 step 1", "parent 1 step 2", "parent 2 step 2"], []],
```
Comment (i-keliukh) [inline _universum/modules/launcher.py]: What is this?

File: _universum/modules/launcher.py [hunk 2]
```diff
@@ -85,6 +85,75 @@ def check_if_env_set(configuration):
     return True
 
 
+def check_str_match(string, expected_substrings, unexpected_substrings):
+    """The function to check whenever specified string contains 'expected' and
```
Comment (k-dovgan) [inline _universum/modules/launcher.py]: 'whether', not 'whenever'

File: _universum/modules/launcher.py [hunk 3]
```diff
@@ -85,6 +85,75 @@ def check_if_env_set(configuration):
...
+    does NOT contains 'unexpected' substrings.
```
Comment (k-dovgan) [inline _universum/modules/launcher.py]: 'does not contain'

File: _universum/modules/launcher.py [hunk 4]
```diff
@@ -85,6 +85,75 @@ def check_if_env_set(configuration):
...
+
+    for substr in unexpected_substrings:
+        if substr in string:
+            result = False
+            break
+    return result
+
+
+def get_match_patterns(filters):
+    """The function to parse 'filters' defined as single string into a lists of
```
Comment (k-dovgan) [inline _universum/modules/launcher.py]: 'as a single string'

Comment (k-dovgan) [inline _universum/modules/launcher.py]: 'into the lists'

File: _universum/modules/launcher.py [hunk 5]
```diff
@@ -85,6 +85,75 @@ def check_if_env_set(configuration):
...
+    >>> get_match_patterns("f:!f 1:f 2:!f 3")
+    ["f", "f 2"], ["f 1", "f 3"]
+    """
+    expected = []
+    unexpected = []
+
+    filters = filters.split(':') if filters else []
+    for f in filters:
+        if f.startswith('!'):
+            unexpected.append(f[1:]) if len(f) > 1 else None # pylint: disable=expression-not-assigned
```
Comment (k-dovgan) [inline _universum/modules/launcher.py]: Do we _really_ need a one-line if in here? Pylint raises a warning here for a reason, a think.

File: _universum/modules/launcher.py [hunk 6]
```diff
@@ -235,6 +304,20 @@ def define_arguments(argument_parser):
         parser.add_argument("--launcher-config-path", "-lcp", dest="config_path", metavar="CONFIG_PATH",
                             help="Project configs.py file location. Mandatory parameter")
 
+        parser.add_argument("--run-step", "-run", dest="run_step",
```
Comment (k-dovgan) [inline _universum/modules/launcher.py]: If this is a launcher parameter, it should start with '--launcher-'; also I'm not sure if 'run step' is clear enough. Could you test this parameter name on someone

File: _universum/modules/launcher.py [hunk 7]
```diff
@@ -235,6 +304,20 @@ def define_arguments(argument_parser):
...
+                            help="Allows to filter which steps to execute during launch. "
```
Comment (k-dovgan) [inline _universum/modules/launcher.py]: Universum has a very, very long help already. I'd like a way shorter help here with link to documentation, describing all this stuff

File: _universum/modules/launcher.py (hunk 5 above)
Comment (i-savynikh) [reply _universum/modules/launcher.py]: we don't want 3-level of nesting

File: _universum/modules/launcher.py [hunk 8]
```diff
@@ -85,6 +85,81 @@ def check_if_env_set(configuration):
...
+
+    for substr in unexpected_substrings:
+        if substr in string:
+            result = False
+            break
+    return result
+
+
+def get_match_patterns(filters):
+    """The function to parse 'filters' defined as a single string into a lists
```
Comment (k-dovgan) [inline _universum/modules/launcher.py]: **`into the lists`**
please do not resolve conversations unless you fix them or comment on why the fix is not needed

File: doc/args.rst [hunk 9]
```diff
@@ -17,6 +17,21 @@ All command-line parameters, general and module-related, are passed to this main
...
         Display product name & version instead of launching.
 
+    --steps-filter -f : @replace
+        | Allows to filter which steps to execute during launch.
+         String value representing single filter or a set of filters separated by '**:**'.
+         To define exclude pattern use '**!**' symbol at the beginning of pattern.
+        |
+        | A universum step match specified pattern when 'filter' is a substring of step 'name'.
+         This functionality is similar to 'boosttest' and 'gtest' filtering.
+         Except using of special characters like '*', '?', etc. which are ignored.
```
Comment (k-dovgan) [inline doc/args.rst]: This functionality is similar to 'boosttest' and 'gtest' filtering,  except special characters (like '*', '?', etc.) are ignored.

File: doc/args.rst [hunk 10]
```diff
@@ -17,6 +17,21 @@ All command-line parameters, general and module-related, are passed to this main
...
+        |
+        | Examples:
+        | * -f='run test'               - run only steps which contain 'run test' string in name
```
Comment (k-dovgan) [inline doc/args.rst]: steps that contain 'run test' string in their names

File: doc/args.rst [hunk 11]
```diff
@@ -17,6 +17,21 @@ All command-line parameters, general and module-related, are passed to this main
...
+        | * -f='!run test'              - run all steps except those which contain 'run test' string in name
```
Comment (k-dovgan) [inline doc/args.rst]: except those containing 'run test' string in their names

File: doc/args.rst [hunk 12]
```diff
@@ -17,6 +17,21 @@ All command-line parameters, general and module-related, are passed to this main
...
+        | * -f='test 1:test 2'          - run all steps with 'test 1' OR 'test 2' in name
```
Comment (k-dovgan) [inline doc/args.rst]: in their names

File: doc/args.rst [hunk 13]
```diff
@@ -17,6 +17,21 @@ All command-line parameters, general and module-related, are passed to this main
...
+        | * -f='test 1:!unit test 1'    - run all steps with 'test 1' in name except those which contain 'unit test 1'
```
Comment (k-dovgan) [inline doc/args.rst]: in their names except those also containing 'unit test 1' substring

File: _universum/modules/launcher.py [hunk 14]
```diff
@@ -278,7 +364,10 @@ def process_project_configs(self):
             dump_file.write(self.source_project_configs.dump())
             dump_file.close()
 
-            self.project_configs = self.source_project_configs.filter(check_if_env_set)
+            configs = self.source_project_configs.filter(check_if_env_set)
+            self.project_configs = configs.filter(lambda config: check_str_match(config['name'],
```
Comment (k-dovgan) [inline _universum/modules/launcher.py]: Do we really need a lambda in here?

File: _universum/modules/launcher.py [hunk 15]
```diff
@@ -226,15 +301,26 @@ def define_arguments(argument_parser):
...
 
-        parser.add_argument("--launcher-config-path", "-lcp", dest="config_path", metavar="CONFIG_PATH",
+        parser.add_argument("--steps-config", "-cfg", dest="config_path", metavar="CONFIG_PATH",
                             help="Project configs.py file location. Mandatory parameter")
 
+        parser.add_argument("--steps-filter", "-f", dest="step_filter", action='append',
+                            help="Filter steps to execute. A single filter or a set of filters separated by ':'. "
+                                 "Exlude using '!' symbol before filter. "
+                                 "Example: -f='str1:!not str2' OR -f='str1' -f='!not str2'. "
+                                 "See online docs for more details. ")
```
Comment (k-dovgan) [inline _universum/modules/launcher.py]: Last whitespace in help is not needed

File: _universum/modules/launcher.py (hunk 14 above)
Comment (i-savynikh) [reply _universum/modules/launcher.py]: yes, we need

**Pull request merged**